atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/InMemoryKeyValueService.java

```diff
@@ -476,7 +478,29 @@ public void checkAndSet(CheckAndSetRequest request) throws CheckAndSetException
         Cell cell = request.cell();
         Optional<byte[]> oldValue = request.oldValue();
         byte[] contents = request.newValue();
+        checkAndSetInternal(tableRef, table, cell, oldValue, contents);
+    }
+
+    @Override
+    public void multiCheckAndSet(MultiCheckAndSetRequest multiCheckAndSetRequest) throws MultiCheckAndSetException {
+        TableReference tableRef = multiCheckAndSetRequest.tableRef();
+        Table table = getTableMap(tableRef);
+
+        multiCheckAndSetRequest.updates().forEach((cell, val) -> {
```

Comment:
This is not atomic. Unfortunately, looks like `ConcurrentSkipListMap` and `ConcurrentHashMap` implementations of `replaceAll` do not give us the concurrency guarantees we need either. I think it's ok to document this for now, and we can potentially fix it later if needed by introducing write locks that this method takes out, but it's a bit awful because every other write method will then need to get read locks. Given that Cassandra also does not do this in a completely atomic way, this is probably fine, but we do need to throw a correct exception when we conflict
@jeremyk-91 if you want to weigh in